builder/job.go

```diff
@@ -36,44 +36,61 @@ var validCommitCommands = map[string]bool{
 	"onbuild":    true,
 }
 
-type BuilderJob struct {
-	Engine *engine.Engine
-	Daemon *daemon.Daemon
+type Config struct {
+	DockerfileName string
+	RemoteURL      string
+	RepoName       string
+	SuppressOutput bool
+	NoCache        bool
+	Remove         bool
+	ForceRemove    bool
+	Pull           bool
+	JSONFormat     bool
+	Memory         int64
+	MemorySwap     int64
+	CpuShares      int64
+	CpuSetCpus     string
+	CpuSetMems     string
+	AuthConfig     *registry.AuthConfig
+	ConfigFile     *registry.ConfigFile
+
+	Stdout  io.Writer
+	Context io.ReadCloser
+	// When closed, the job has been cancelled.
+	// Note: not all jobs implement cancellation.
+	// See Job.Cancel() and Job.WaitCancelled()
+	cancelled  chan struct{}
+	cancelOnce sync.Once
 }
 
-func (b *BuilderJob) Install() {
-	b.Engine.Register("build", b.CmdBuild)
-	b.Engine.Register("build_config", b.CmdBuildConfig)
+// When called, causes the Job.WaitCancelled channel to unblock.
+func (b *Config) Cancel() {
+	b.cancelOnce.Do(func() {
```

Comment:
You know, you can do it like
```
select {
case <-cancelled:
default:
    close(cancelled)
}
```
But I personally don't care, once is ok too. Just shared trick :)